File: platform/vapi/mock-server/fixtures.ts
/**
 * Canned tool responses for local development.
 *
 * Each returns a SPOKEN SENTENCE — the same contract Core API will honour (doc 04 §5.1):
 * numbers in spoken form, at most three options, machine data only as an echo-able token.
 *
 * The clock is frozen via GRACE_MOCK_NOW so "Tuesday the fourth" is reproducible across
 * runs, which is what makes voice simulations deterministic.
 */
import type {
  CancelAppointmentInput,
  CheckAvailabilityInput,
  CreateBookingInput,
  GetBusinessInfoInput,
  GetServicesAndPricingInput,
  LookupCustomerInput,
  RescheduleAppointmentInput,
  TakeMessageInput,
} from '@grace/contracts';

import { chicagoDate, speakDate, speakList, speakPrice, speakTime } from './speech.js';

export function now(): Date {
  const override = process.env['GRACE_MOCK_NOW'];
  return override ? new Date(override) : new Date();
}

// ── seed data ──────────────────────────────────────────────────────────────────

const SERVICES = [
  { serviceCode: 'massage_60', name: '60-minute massage', durationMinutes: 60, priceCents: 13500, memberPriceCents: 11500, depositRequired: false },
  { serviceCode: 'massage_90', name: '90-minute massage', durationMinutes: 90, priceCents: 18500, memberPriceCents: 16000, depositRequired: true },
  { serviceCode: 'deep_tissue_60', name: '60-minute deep tissue', durationMinutes: 60, priceCents: 15000, memberPriceCents: 13000, depositRequired: false },
];

const PROVIDERS = [
  { id: 'prv_maria', name: 'Maria' },
  { id: 'prv_james', name: 'James' },
];

const BUSINESS_INFO: Record<GetBusinessInfoInput['topic'], string> = {
  hours: 'We are open Monday through Saturday, nine in the morning until seven in the evening, and closed Sundays.',
  location: 'We are on Dundee Road in Buffalo Grove, just past the Town Center, with parking right out front.',
  parking: 'There is free parking directly in front of the building.',
  contact: 'The best way is right here on this line, or by text to the same number.',
  services_overview: 'We offer therapeutic massage, deep tissue, acupuncture, and cryotherapy.',
  policies: 'Changes and cancellations are free up to forty-eight hours before your appointment.',
  memberships: 'Members get a reduced rate on every service and priority booking.',
};

/** Stable pseudo-random so the same call id yields the same slots. */
function hash(s: string): number {
  let h = 2166136261;
  for (let i = 0; i < s.length; i++) {
    h ^= s.charCodeAt(i);
    h = Math.imul(h, 16777619);
  }
  return Math.abs(h);
}

const SLOT_ALPHABET = '0123456789ABCDEFGHJKMNPQRSTVWXYZ';
function slotId(seed: string): string {
  const h = hash(seed);
  let out = '';
  for (let i = 0; i < 3; i++) out += SLOT_ALPHABET[(h >> (i * 5)) % SLOT_ALPHABET.length] ?? '0';
  return `hold-${out}`;
}

/** In-memory bookings so createBooking → cancel/reschedule behaves coherently in one session. */
const bookings = new Map<string, { ref: string; startsAt: string; provider: string; service: string }>();

export function reset(): void {
  bookings.clear();
}

// ── handlers ───────────────────────────────────────────────────────────────────

export const FIXTURES = {
  getBusinessInfo: (args: GetBusinessInfoInput): string => {
    return BUSINESS_INFO[args.topic];
  },

  lookupCustomer: (args: LookupCustomerInput): string => {
    // Even digits → a known member, so both branches are reachable deterministically.
    const known = hash(args.phone) % 2 === 0;
    return known
      ? 'That number matches Jordan, who has been in four times and is a member.'
      : 'I do not see that number on file yet.';
  },

  getServicesAndPricing: (args: GetServicesAndPricingInput): string => {
    const q = args.query.toLowerCase();
    const matches = SERVICES.filter(
      (s) =>
        s.name.toLowerCase().includes(q) ||
        q.split(/\s+/).some((w) => w.length > 3 && s.name.toLowerCase().includes(w)),
    );
    const picked = (matches.length > 0 ? matches : SERVICES).slice(0, 3);
    const phrases = picked.map((s) => {
      const cents = args.isMember ? s.memberPriceCents : s.priceCents;
      return `the ${s.name} at ${speakPrice(cents)}`;
    });
    return `We have ${speakList(phrases)}.`;
  },

  checkAvailability: (args: CheckAvailabilityInput, callId: string): string => {
    const startHour = args.timePreference === 'morning' ? 9 : args.timePreference === 'evening' ? 17 : 13;

    const slots = [0, 1, 2].map((i) => {
      // chicagoDate, not setHours — the model's date is a Chicago calendar date and the
      // process timezone must not leak into it.
      const d = chicagoDate(args.preferredDate, startHour + i, i === 1 ? 30 : 15);
      const provider = PROVIDERS[i % PROVIDERS.length];
      return {
        id: slotId(`${callId}:${args.preferredDate}:${String(i)}`),
        startsAt: d.toISOString(),
        provider: provider?.name ?? 'Maria',
      };
    });

    const wanted = args.providerPreference?.toLowerCase();
    const filtered = wanted ? slots.filter((s) => s.provider.toLowerCase() === wanted) : slots;
    const offer = (filtered.length > 0 ? filtered : slots).slice(0, 3);

    if (offer.length === 0) {
      return `I don't have anything on ${speakDate(args.preferredDate)}. Would another day work?`;
    }

    const phrases = offer.map((s) => `${speakTime(s.startsAt)} with ${s.provider}`);
    const prefix =
      wanted && filtered.length === 0
        ? `${String(args.providerPreference)} isn't available then, but I have `
        : `I have `;
    return `${prefix}${speakList(phrases)} on ${speakDate(args.preferredDate)}. Which works?`;
  },

  createBooking: (args: CreateBookingInput, callId: string): string => {
    // Server-side medical gate: the prompt is not the only thing enforcing this (I4).
    if (!args.medicalScreenPassed) {
      return "Before I book, I'd like one of our team to go over a couple of health questions with you.";
    }
    const ref = `bk-${slotId(`${callId}:${args.slotId}`).slice(5)}9`;
    const startsAt = new Date(now().getTime() + 86_400_000).toISOString();
    bookings.set(ref.toLowerCase(), { ref, startsAt, provider: 'Maria', service: '60-minute massage' });
    return `You're all set, ${args.firstName} — ${speakDate(startsAt)} at ${speakTime(startsAt)} with Maria. I'll text you a confirmation.`;
  },

  rescheduleAppointment: (args: RescheduleAppointmentInput): string => {
    const b = bookings.get(args.bookingRef.toLowerCase());
    if (!b) return "I can't find that appointment — let me get someone who can look it up properly.";
    const startsAt = new Date(`${args.newDate}T18:30:00-05:00`).toISOString();
    return `Done — I've moved you to ${speakDate(startsAt)} at ${speakTime(startsAt)}. There's no charge for that change.`;
  },

  cancelAppointment: (args: CancelAppointmentInput): string => {
    const b = bookings.get(args.bookingRef.toLowerCase());
    if (!b) return "I can't find that appointment — let me get someone who can look it up properly.";
    bookings.delete(args.bookingRef.toLowerCase());
    return `That's cancelled for you. Since it's more than forty-eight hours away, there's no charge.`;
  },

  // Async tools: Vapi never delivers these to the model, so the string is for logs only.
  sendIntakeForm: (): string => 'intake form queued',
  sendDepositLink: (): string => 'deposit link queued',
  sendBookingConfirmation: (): string => 'confirmation queued',

  takeMessage: (args: TakeMessageInput): string => {
    return `Got it — I've passed that to the team and someone will call you back about ${args.subject.toLowerCase()}.`;
  },

  flagMedicalHold: (): string => {
    return "Thanks for telling me — I'd like one of our team to go over that with you before we book.";
  },

  flagEscalation: (): string => 'escalation logged',
} as const;

File: src/grace_platform/vapi/mock_server/fixtures.py
"""Canned tool responses for local development.

Each returns a SPOKEN SENTENCE — the same contract Core API will honour (doc 04 §5.1):
numbers in spoken form, at most three options, machine data only as an echo-able token.

The clock is frozen via GRACE_MOCK_NOW so dates are reproducible, which is what makes
voice simulations deterministic.
"""

from __future__ import annotations

import os
from collections.abc import Callable
from datetime import UTC, datetime, timedelta
from typing import Any, NamedTuple

from .speech import chicago_date, speak_date, speak_list, speak_price, speak_time


def now() -> datetime:
    override = os.environ.get("GRACE_MOCK_NOW")
    return (
        datetime.fromisoformat(override.replace("Z", "+00:00")) if override else datetime.now(UTC)
    )


class Service(NamedTuple):
    code: str
    name: str
    price: int
    member: int


SERVICES = (
    Service("massage_60", "60-minute massage", 13500, 11500),
    Service("massage_90", "90-minute massage", 18500, 16000),
    Service("deep_tissue_60", "60-minute deep tissue", 15000, 13000),
)

PROVIDERS = ("Maria", "James")

BUSINESS_INFO = {
    "hours": "We are open Monday through Saturday, nine in the morning until seven in the evening, and closed Sundays.",
    "location": "We are on Dundee Road in Buffalo Grove, just past the Town Center, with parking right out front.",
    "parking": "There is free parking directly in front of the building.",
    "contact": "The best way is right here on this line, or by text to the same number.",
    "services_overview": "We offer therapeutic massage, deep tissue, acupuncture, and cryotherapy.",
    "policies": "Changes and cancellations are free up to forty-eight hours before your appointment.",
    "memberships": "Members get a reduced rate on every service and priority booking.",
}

_ALPHABET = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"


def _hash(s: str) -> int:
    h = 2166136261
    for ch in s:
        h = ((h ^ ord(ch)) * 16777619) & 0xFFFFFFFF
    return h


def _slot_id(seed: str) -> str:
    h = _hash(seed)
    return "hold-" + "".join(_ALPHABET[(h >> (i * 5)) % len(_ALPHABET)] for i in range(3))


_bookings: dict[str, dict[str, Any]] = {}


def reset() -> None:
    _bookings.clear()


def get_business_info(a: dict[str, Any], _call: str) -> str:
    return BUSINESS_INFO[a["topic"]]


def lookup_customer(a: dict[str, Any], _call: str) -> str:
    # Even hash → a known member, so both branches are reachable deterministically.
    if _hash(a["phone"]) % 2 == 0:
        return "That number matches Jordan, who has been in four times and is a member."
    return "I do not see that number on file yet."


def get_services_and_pricing(a: dict[str, Any], _call: str) -> str:
    q = a["query"].lower()
    matches = [s for s in SERVICES if q in s.name.lower()] or list(SERVICES)
    member = bool(a.get("isMember"))
    phrases = [
        f"the {s.name} at {speak_price(s.member if member else s.price)}" for s in matches[:3]
    ]
    return f"We have {speak_list(phrases)}."


def check_availability(a: dict[str, Any], call_id: str) -> str:
    pref = a.get("timePreference", "any")
    start = 9 if pref == "morning" else 17 if pref == "evening" else 13
    date = a["preferredDate"]

    slots: list[tuple[datetime, str]] = [
        (chicago_date(date, start + i, 30 if i == 1 else 15), PROVIDERS[i % len(PROVIDERS)])
        for i in range(3)
    ]

    wanted = str(a.get("providerPreference") or "").lower()
    filtered = [s for s in slots if s[1].lower() == wanted] if wanted else slots
    offer = (filtered or slots)[:3]

    phrases = [f"{speak_time(starts)} with {provider}" for starts, provider in offer]
    prefix = (
        f"{a['providerPreference']} isn't available then, but I have "
        if wanted and not filtered
        else "I have "
    )
    return f"{prefix}{speak_list(phrases)} on {speak_date(date)}. Which works?"


def create_booking(a: dict[str, Any], call_id: str) -> str:
    # Server-side medical gate: the prompt is not the only thing enforcing this (I4).
    if not a.get("medicalScreenPassed"):
        return "Before I book, I'd like one of our team to go over a couple of health questions with you."
    ref = "bk-" + _slot_id(f"{call_id}:{a['slotId']}")[5:] + "9"
    starts = now() + timedelta(days=1)
    _bookings[ref.lower()] = {"ref": ref, "starts": starts}
    return (
        f"You're all set, {a['firstName']} — {speak_date(starts)} at {speak_time(starts)} "
        f"with Maria. I'll text you a confirmation."
    )


def reschedule_appointment(a: dict[str, Any], _call: str) -> str:
    if a["bookingRef"].lower() not in _bookings:
        return "I can't find that appointment — let me get someone who can look it up properly."
    starts = chicago_date(a["newDate"], 18, 30)
    return (
        f"Done — I've moved you to {speak_date(starts)} at {speak_time(starts)}. "
        f"There's no charge for that change."
    )


def cancel_appointment(a: dict[str, Any], _call: str) -> str:
    if _bookings.pop(a["bookingRef"].lower(), None) is None:
        return "I can't find that appointment — let me get someone who can look it up properly."
    return (
        "That's cancelled for you. Since it's more than forty-eight hours away, there's no charge."
    )


def take_message(a: dict[str, Any], _call: str) -> str:
    return (
        f"Got it — I've passed that to the team and someone will call you back about "
        f"{a['subject'].lower()}."
    )


def flag_medical_hold(_a: dict[str, Any], _call: str) -> str:
    return (
        "Thanks for telling me — I'd like one of our team to go over that with you before we book."
    )


Fixture = Callable[[dict[str, Any], str], str]

# Async tools: Vapi never delivers these to the model, so the string is for logs only.
FIXTURES: dict[str, Fixture] = {
    "getBusinessInfo": get_business_info,
    "lookupCustomer": lookup_customer,
    "getServicesAndPricing": get_services_and_pricing,
    "checkAvailability": check_availability,
    "createBooking": create_booking,
    "rescheduleAppointment": reschedule_appointment,
    "cancelAppointment": cancel_appointment,
    "sendIntakeForm": lambda _a, _c: "intake form queued",
    "sendDepositLink": lambda _a, _c: "deposit link queued",
    "sendBookingConfirmation": lambda _a, _c: "confirmation queued",
    "takeMessage": take_message,
    "flagMedicalHold": flag_medical_hold,
    "flagEscalation": lambda _a, _c: "escalation logged",
}
